provd/internal/testutils/gdm_mock.go

```diff
@@ -0,0 +1,222 @@
+	return nil
+}
+
+func (g *gdmSessionBus) StartSessionWhenReady(serviceName string, shouldStartSession bool) *dbus.Error {
+	return nil
+}
+
+// emitSignal is essentially a copy of dbus.Conn.Emit, but with the ability to specify the destination.
+func emitSignal(conn *dbus.Conn, path dbus.ObjectPath, destination, name string, values ...interface{}) error {
+	i := strings.LastIndex(name, ".")
+	iface := name[:i]
+	member := name[i+1:]
+	msg := new(dbus.Message)
+	msg.Type = dbus.TypeSignal
+	msg.Headers = make(map[dbus.HeaderField]dbus.Variant)
+	msg.Headers[dbus.FieldInterface] = dbus.MakeVariant(iface)
+	msg.Headers[dbus.FieldMember] = dbus.MakeVariant(member)
+	msg.Headers[dbus.FieldPath] = dbus.MakeVariant(path)
+	msg.Headers[dbus.FieldDestination] = dbus.MakeVariant(destination)
+	msg.Body = values
+	if len(values) > 0 {
+		msg.Headers[dbus.FieldSignature] = dbus.MakeVariant(dbus.SignatureOf(values...))
+	}
+	if err := msg.IsValid(); err != nil {
+		return err
+	}
+	return conn.Send(msg, nil).Err
+}
+
+func setupPrivateBus() (*privateBus, error) {
+	address, tempDir, cleanup := startPrivateDbusServer()
+	conn, err := dbus.Dial(address)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to session bus: %w", err)
+	}
+	err = conn.Auth(nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to authenticate: %w", err)
+	}
+	err = conn.Hello()
+	if err != nil {
+		return nil, fmt.Errorf("failed to say hello: %w", err)
+	}
+	g := &gdmSessionBus{conn: conn}
+	if err := conn.Export(g, dbus.ObjectPath("/org/gnome/DisplayManager/Session"), consts.DbusGdmPrefix+".UserVerifier"); err != nil {
+		return nil, fmt.Errorf("could not export UserVerifier mock: %w", err)
+	}
+	if err := conn.Export(g, dbus.ObjectPath("/org/gnome/DisplayManager/Session"), consts.DbusGdmPrefix+".Greeter"); err != nil {
+		return nil, fmt.Errorf("could not export Greeter mock: %w", err)
+	}
+	if err := conn.Export(g, dbus.ObjectPath("/org/gnome/DisplayManager/Session"), consts.DbusPeerPrefix); err != nil {
+		return nil, fmt.Errorf("could not export Peer mock: %w", err)
+	}
+
+	var reply dbus.RequestNameReply
+	reply, err = conn.RequestName(consts.DbusGdmPrefix+".UserVerifier", dbus.NameFlagDoNotQueue)
+	if err != nil {
+		return nil, fmt.Errorf("failed to acquire account name on local system bus: %w", err)
+	}
+	if reply != dbus.RequestNameReplyPrimaryOwner {
+		return nil, fmt.Errorf("failed to acquire account name on local system bus: name already taken")
+	}
+
+	reply, err = conn.RequestName(consts.DbusGdmPrefix+".Greeter", dbus.NameFlagDoNotQueue)
+	if err != nil {
+		return nil, fmt.Errorf("failed to acquire account name on local system bus: %w", err)
+	}
+	if reply != dbus.RequestNameReplyPrimaryOwner {
+		return nil, fmt.Errorf("failed to acquire account name on local system bus: name already taken")
+	}
+
+	return &privateBus{conn, tempDir, address, cleanup}, nil
+}
+
+func startPrivateDbusServer() (string, string, func()) {
```

Comment:
I've mostly re-used the code from `systembus.go` for this function, but kept it separate for now. It might make sense generalize `systembus.go` and move the private bus stuff there too at some point.

Reply:
Fair enough, but agree its fine to have a little repetition for now :slightly_smiling_face: